web/hobbit-ackinfo.cgi.1
.TH HOBBIT-ACKINFO.CGI 1 "Version 4.2.2: 15 Dec 2008" "Xymon"
.SH NAME
hobbit-ackinfo.cgi \- Xymon CGI script to acknowledge alerts
.SH SYNOPSIS
.B "hobbit-ackinfo.cgi"

.SH DESCRIPTION
\fBhobbit-ackinfo.cgi\fR is invoked as a CGI script via the 
hobbit-ackinfo.sh CGI wrapper.

hobbit-ackinfo.cgi is used to acknowledge an alert on the Xymon
"Critical Systems" view, generated by the
.I hobbit-nkview.cgi(1)
utility. This allows the staff viewing the Critical Systems view to 
acknowledge alerts with a "Level 1" alert, thereby removing the
alert from the Critical Systems view.

Note that the Level 1 alert generated by the hobbit-ackinfo.cgi
utility does \fBNOT\fR stop alerts from being sent.

In a future version of Xymon (after Xymon 4.2), this utility
will also be used for acknowledging alerts at other levels.

.SH OPTIONS
.IP "--level=NUMBER"
Sets the acknowledgment level. This is typically used to force a specific
level of the acknowledgment, e.g. a level 1 acknowledge when called from the
Critical Systems view.

.IP "--validity=TIME"
Sets the validity of the acknowledgment. By default this is taken from the
CGI parameters supplied by the user.

.IP "--sender=STRING"
Logs STRING as the sender of the acknowledgment. By default, this is taken
from the loginname of the webuser sending the acknowledgment.

.IP "--env=FILENAME"
Loads the environment defined in FILENAME before executing the CGI script.

.IP "--area=NAME"
Load environment variables for a specific area. NB: if used,
this option must appear before any --env=FILENAME option.

.IP "--debug"
Enables debugging output.

.SH "ENVIRONMENT VARIABLES"
.IP BBHOME
Used to locate the template files for the generated web pages.

.IP QUERY_STRING
Contains the parameters for the CGI script.

.SH "SEE ALSO"
hobbit-nkview.cgi(1), xymon(7)



Changes
Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------
r5940 | storner | 2008-11-26 13:17:12 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitfetch.c

hobbitfetch insisted that hosts must have a valid IP-adress in bb-hosts. This causes problems for hosts with dynamic IP-adresses. This patch causes hobbitfetch to lookup the IP-address at run-time if the IP is listed as "0.0.0.0".


------------------------------------------------------------------------
r5941 | storner | 2008-11-26 13:20:44 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbitsvc-info.c

Hostname filtering on the info column page is broken, if you have multiple hosts with similar names, resulting in the same columns showing up multiple times in the disable section. A similar problem affects host filtering in the configuration report tool.


------------------------------------------------------------------------
r5942 | storner | 2008-11-26 13:21:27 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris client would not collect the "top" process-listing data which is normally included in the data on the "cpu" column.


------------------------------------------------------------------------
r5943 | storner | 2008-11-26 13:25:28 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/build/bb-commands.sh
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris lofs (loopback filesystem) should not be included in the df output appearing on the Disk status.
------------------------------------------------------------------------
r5944 | storner | 2008-11-26 13:29:56 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/lib/cgi.c
   M /branches/4.2.0/lib/cgi.h
   M /branches/4.2.0/web/bb-ack.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-hostgraphs.c
   M /branches/4.2.0/web/hobbit-statusreport.c

Cookie handling does not always work. This results in hosts not showing up on the new acknowledgment page, or hosts missing from the Metrics report.
------------------------------------------------------------------------
r5945 | storner | 2008-11-26 13:30:31 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/bbdisplay/bbgen.h

Hobbit 4.2.0 fails to build on some OS X releases due to a naming conflict between a system-defined datatype, and a different datatype defined in Hobbit.
------------------------------------------------------------------------
r5946 | storner | 2008-11-26 13:31:41 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/zoom.js
   M /branches/4.2.0/lib/hobbitrrd.c
   M /branches/4.2.0/lib/hobbitrrd.h
   M /branches/4.2.0/lib/htmllog.c
   M /branches/4.2.0/web/hobbitgraph.c
   M /branches/4.2.0/web/hobbitsvc-trends.c

When viewing detailed graphs linked from the detailed status page, the background color of the webpages would always show green, regardless of the actual color of the status it was linked from. This patch keeps the background color at what the status page shows.
------------------------------------------------------------------------
r5947 | storner | 2008-11-26 13:32:55 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/confreport_front
   M /branches/4.2.0/hobbitd/wwwfiles/menu/menu_items.js.DIST
   M /branches/4.2.0/web/Makefile
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-confreport.cgi.1

The Configuration Report did not include information about what systems appears on the "Critical Systems" view.

web/hobbit-ackinfo.c
/*----------------------------------------------------------------------------*/
/* Hobbit CGI for sending in an "ackinfo" message.                            */
/*                                                                            */
/* Copyright (C) 2005-2008 Henrik Storner <henrik@storner.dk>                 */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id: hobbit-ackinfo.c,v 1.5 2008/01/03 10:04:58 henrik Exp henrik $";

#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <limits.h>
#include <ctype.h>

#include "libbbgen.h"

static char *hostname = NULL;
static char *testname = NULL;
static int level = -1;
static int validity = -1;
static char *ackedby = NULL;
static char *ackmsg  = NULL;

static void parse_query(void)
{
	cgidata_t *cgidata = cgi_request();
	cgidata_t *cwalk;

	cwalk = cgidata;
	while (cwalk) {
		if (strcasecmp(cwalk->name, "HOST") == 0) {
			hostname = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "SERVICE") == 0) {
			testname = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "ALLTESTS") == 0) {
			if (strcasecmp(cwalk->value, "on") == 0) testname = strdup("*");
		}
		else if (strcasecmp(cwalk->name, "NOTE") == 0) {
			ackmsg = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "LEVEL") == 0) {
			/* Commandline may override this */
			if (level == -1) level = atoi(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "VALIDITY") == 0) {
			/* Commandline may override this */
			if (validity == -1) validity = atoi(cwalk->value);
		}

		cwalk = cwalk->next;
	}
}


int main(int argc, char *argv[])
{
	int argi;
	char *envarea = NULL;
	char *bbmsg;
	int res;

	for (argi = 1; (argi < argc); argi++) {
		if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--level=")) {
			char *p = strchr(argv[argi], '=');
			level = atoi(p+1);
		}
		else if (argnmatch(argv[argi], "--validity=")) {
			char *p = strchr(argv[argi], '=');
			validity = atoi(p+1);
		}
		else if (argnmatch(argv[argi], "--sender=")) {
			char *p = strchr(argv[argi], '=');
			ackedby = strdup(p+1);
		}
	}

	redirect_cgilog("hobbit-ackinfo");
	parse_query();

	if (hostname && *hostname && testname && *testname && ((level == 0) || (validity>0)) && ackmsg && *ackmsg) {
		char *p;

		/* Get the login username */
		if (!ackedby) ackedby = getenv("REMOTE_USER");
		if (!ackedby) ackedby = "UnknownUser";

		if (validity == -1) validity = 30; /* 30 minutes */
		validity = validity*60;

		p = strchr(ackmsg, '\n'); if (p) *p = '\0';

		/* ackinfo HOST.TEST\nlevel\nvaliduntil\nackedby\nmsg */
		bbmsg = (char *)malloc(1024 + strlen(hostname) + strlen(testname) + strlen(ackedby) + strlen(ackmsg));
		sprintf(bbmsg, "ackinfo %s.%s\n%d\n%d\n%s\n%s\n",
			hostname, testname, level, validity, ackedby, ackmsg);
		res = sendmessage(bbmsg, NULL, BBTALK_TIMEOUT, NULL);
	}
	else {
		bbmsg = (char *)malloc(4096);
		sprintf(bbmsg, "error in input params: hostname=%s, testname=%s, ackmsg=%s, validity=%d\n",
			hostname, testname, ackmsg, validity);
	}

	fprintf(stdout, "Content-type: %s\n", xgetenv("HTMLCONTENTTYPE"));
	fprintf(stdout, "Location: %s\n", getenv("HTTP_REFERER"));
	fprintf(stdout, "\n");
	fprintf(stdout, "Sent to hobbitd:\n%s\n", bbmsg);

	return 0;
}

